lndsigner.go

```diff
@@ -0,0 +1,340 @@
+	// creating the signer macaroon file and the accounts list file.
+	//
+	// Why 640 is safe:
+	// Assuming a reasonably secure Linux system, it will have a
+	// separate group for each user. E.g. a new user lnd gets assigned group
+	// lnd which nothing else belongs to. A system that does not do this is
+	// inherently broken already.
+	//
+	// Since there is no other user in the group, no other user can read
+	// admin macaroon unless the administrator explicitly allowed it. Thus
+	// there's no harm allowing group read.
+	outputFilePermissions = 0640
+)
+
+// ListenerWithSignal is a net.Listener that has an additional Ready channel
+// that will be closed when a server starts listening.
+type ListenerWithSignal struct {
+	net.Listener
+
+	// Ready will be closed by the server listening on Listener.
+	Ready chan struct{}
+
+	// MacChan is an optional way to pass the admin macaroon to the program
+	// that started lnd. The channel should be buffered to avoid lnd being
+	// blocked on sending to the channel.
+	MacChan chan []byte
+}
+
+// ListenerCfg is a wrapper around custom listeners that can be passed to lnd
+// when calling its main method.
+type ListenerCfg struct {
+	// RPCListeners can be set to the listeners to use for the RPC server.
+	// If empty a regular network listener will be created.
+	RPCListeners []*ListenerWithSignal
+}
+
+// Main is the true entry point for lnd. It accepts a fully populated and
+// validated main configuration struct and an optional listener config struct.
+// This function starts all main system components then blocks until a signal
+// is received on the shutdownChan at which point everything is shut down again.
+func Main(cfg *Config, lisCfg ListenerCfg) error {
+	// mkErr makes it easy to return logged errors.
+	mkErr := func(format string, args ...interface{}) error {
+		signerLog.Errorf("Shutting down because error in main "+
+			"method: "+format, args...)
+		return fmt.Errorf(format, args...)
+	}
+
+	var network string
+	switch {
+	/*case cfg.MainNet:
+	network = "mainnet"
+	*/
+	case cfg.TestNet3:
+		network = "testnet"
+
+	case cfg.SimNet:
+		network = "simnet"
+
+	case cfg.RegTest:
+		network = "regtest"
+
+	case cfg.SigNet:
+		network = "signet"
+	}
+
+	signerLog.Infof("Active chain: %v (network=%v)", "bitcoin", network)
+
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Use defaults for vault client, including getting config from env.
+	vaultClient, err := api.NewClient(nil)
+	if err != nil {
+		return mkErr("error creating vault client: %v", err)
+	}
+
+	signerClient := vaultClient.Logical()
+
+	listAcctsResp, err := signerClient.ReadWithData(
+		"lndsigner/lnd-nodes/accounts",
+		map[string][]string{
+			"node": []string{cfg.NodePubKey},
+		},
+	)
+	if err != nil {
+		return mkErr("error listing accounts: %v", err)
+	}
+
+	acctList, ok := listAcctsResp.Data["acctList"]
+	if !ok {
+		return mkErr("accounts not returned")
+	}
+
+	err = os.WriteFile(
```

Comment:
Maybe this writing should be a separate command-line command? It's also a question how this is going to work in an actual deployment where you don't have access to the local filesystem. Perhaps an alternative is an extra grpc endpoint to get the accounts. Or just don't do it at all and let users make a call to vault directly to get the accounts.json?

Reply:
Ties into the key import feature as well.

Reply:
I've removed this altogether and described how to get the account JSON directly from `vault` in the README.md. For key import, once the key is imported, it will be possible to retrieve the account list in the same way (I haven't yet implemented import though).